### ISSUE TITLE
Python crash during backward

### ISSUE THREAD

**Issue by EtienneDesticourt** [pytorch/pytorch#2718]
Hi,

I'm running the windows version (torch==0.2.1 +a4fc05a) so if you're not interested in its issues, and if it has no bearing on the other versions, feel free to disregard this issue.

I was trying to build an adversarial autoencoder with multiple discriminators and python crashes in the middle of writing a Traceback when I call backward with this architecture. It also starts consuming huge amounts of memory and the cpu goes full throttle. I think it might have to do with the way I slice some variables.

It might be related to issue #1318, I don't know enough about the backend to be sure so I'm opening this new issue.

Here's a minimal example to reproduce the issue:

```python
import torch
import torch.nn as nn
from torch.autograd import Variable
import numpy as np

NUM_CLASSES = 4
BATCH_SIZE = 256
ENCODER_INPUT_DIM = 200
LATENT_DIM = 2
OUTPUT_DIM = 1
HIDDEN_DIM = 512
EPSILON = 1e-8

encoder = nn.Sequential(
    nn.Linear(ENCODER_INPUT_DIM, HIDDEN_DIM),
    nn.Linear(HIDDEN_DIM, HIDDEN_DIM),
    nn.Linear(HIDDEN_DIM, LATENT_DIM)).cuda()

enc_loss = lambda x: -torch.mean(torch.log(x + EPSILON))

# We create one discriminator per label
discriminators = []
for i in range(NUM_CLASSES):
    discriminator = nn.Sequential(
        nn.Linear(LATENT_DIM, HIDDEN_DIM),
        nn.Linear(HIDDEN_DIM, HIDDEN_DIM),
        nn.Linear(HIDDEN_DIM, OUTPUT_DIM)).cuda()
    discriminators.append(discriminator)

disc_loss = lambda real, fake: -torch.mean(torch.log(real + EPSILON) + torch.log(1 - fake + EPSILON))

batch_x = Variable(torch.randn((BATCH_SIZE, ENCODER_INPUT_DIM))).cuda()
batch_y = Variable(torch.from_numpy(np.random.randint(0, NUM_CLASSES, (BATCH_SIZE,)))).cuda()

generated_latent = encoder(batch_x)
prior_latent = Variable(torch.randn((BATCH_SIZE, LATENT_DIM))).cuda()
for i, disc in enumerate(discriminators):
    # For each discriminator we get the batch data that matches the discriminator's label
    indices = (batch_y==i).data.nonzero()[:, 0].cuda()
    generated = generated_latent[indices, :]
    prior = prior_latent[indices, :]
    
    fake = disc(generated)
    real = disc(prior)
    loss = disc_loss(real, fake)
    loss.backward() # CRASHES HERE ON SECOND ITERATION
    print(i)

```

If we replace:
`generated_latent = encoder(batch_x)`
with
`generated_latent = Variable(encoder(batch_x).data)`
The problem no longer appears, so I'm thinking it has trouble backpropagating the sliced variables through the encoder.

**Comment by soumith** [comment]
thanks for the issue, we'll take a look.

**Comment by soumith** [comment]
took some time to get to this, but once i took a look it was a simple user-error.

So, `generated_latent` is done outside the for-loop.
And in the for-loop, there's this line:
```
generated = generated_latent[indices, :]
```

So now `generated` is connected to `generated_latent`, which means, gradients coming from `generated` will be passed back through to `generated_latent`.

However, once you do `loss.backward()` the graph's temporary buffers are freed by default, which means `generated_latent` cannot backpropagate gradients correctly.

If your intention is not backprop back through `generated_latent` for the discriminators for-loop, then you can simply do:

```python
# rewrap Variable to not back-propagate through generated_latent
generated = Variable(generated_latent[indices, :].data)
```

If your intention is to backprop back through `generated_latent`, then you can simply do:

```python
loss.backward(create_graph=True)
```

**Comment by EtienneDesticourt** [comment]
Yes, this was my intuition as well. But the segmentation fault is the issue, no user error should result in a segmentation fault.
This is not solved, IMO.

**Comment by soumith** [comment]
oh the segmentation fault should be fixed in master. 
whenever @peterjc123 builds another Windows build, it should be fixed for you.

**Comment by EtienneDesticourt** [comment]
Oh awesome then, thanks =)!

**Comment by peterjc123** [comment]
I'm sorry to tell that this problem is not fixed using the current master branch.

**Comment by peterjc123** [comment]
The errors can't be reproduced using the test cases in #1318, so i guess the reasons that cause these two problems are different.

**Comment by soumith** [comment]
@peterjc123 what is a stack-trace for this segfault? I cant reproduce this crash/segfault in Linux / OSX

**Comment by peterjc123** [comment]
The stack trace is listed as the following:
```Traceback
_C.cp36-win_amd64.pyd!torch::autograd::Engine::thread_main(torch::autograd::GraphTask * graph_task) Line 141	C++	Symbols loaded.
[Inline Frame] _C.cp36-win_amd64.pyd!torch::autograd::Engine::thread_init(int) Line 121	C++	Symbols loaded.
_C.cp36-win_amd64.pyd!torch::autograd::python::PythonEngine::thread_init(int device) Line 28	C++	Symbols loaded.
[Inline Frame] _C.cp36-win_amd64.pyd!std::_Invoker_pmf_pointer::_Call(void(torch::autograd::Engine::*)(int) _Pmf, torch::autograd::Engine * &&) Line 1790	C++	Symbols loaded.
[Inline Frame] _C.cp36-win_amd64.pyd!std::invoke(void(torch::autograd::Engine::*)(int) &&) Line 1790	C++	Symbols loaded.
[Inline Frame] _C.cp36-win_amd64.pyd!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int>,std::default_delete<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int> > > >::_Execute(std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int> &) Line 240	C++	Symbols loaded.
[Inline Frame] _C.cp36-win_amd64.pyd!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int>,std::default_delete<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int> > > >::_Run(std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int>,std::default_delete<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int> > > > *) Line 247	C++	Symbols loaded.
[Inline Frame] _C.cp36-win_amd64.pyd!std::_LaunchPad<std::unique_ptr<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int>,std::default_delete<std::tuple<void (__cdecl torch::autograd::Engine::*)(int),torch::autograd::Engine *,int> > > >::_Go() Line 232	C++	Symbols loaded.
_C.cp36-win_amd64.pyd!std::_Pad::_Call_func(void * _Data) Line 210	C++	Symbols loaded.
```
It stucks in the while loop in the function thread_main of engine.cpp. The exact location that it get freezed is:
<strong>
```CPP
FunctionTask task = queue->pop_back();
```
</strong>
At that time, the queue is empty. It happens on two background threads, while the main thread is waiting for the two threads to terminate to get the traceback.

**Comment by peterjc123** [comment]
Resolved in latest builds.
